crates/ruff_python_semantic/src/context.rs

```diff
@@ -18,6 +18,14 @@ use crate::binding::{
 use crate::node::{NodeId, Nodes};
 use crate::scope::{Scope, ScopeId, ScopeKind, Scopes};
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Snapshot {
+    scope_id: ScopeId,
+    stmt_id: Option<NodeId>,
+    in_annotation: bool,
+    in_type_checking_block: bool,
+}
```

Comment:
This is a combination of what was previously called `deferred::Context` and `ast::AnnotationContext`. We're now storing _more_ data than before, since we're now storing `in_type_checking_block` and `in_annotation` for all deferrals, whereas before, we only stored these for deferred type definitions. But I think it's wrong _not_ to snapshot and restore these.